### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by notelevision99**

### CODE REVIEW

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application.Contracts/Samples/BookForList.cs [hunk 1]
```diff
@@ -2,21 +2,21 @@
 using System.Collections.Generic;
 using System.Text;
 
-namespace Acme.ProjectCompare.Samples
-{
-    public class BookForList
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application.Contracts/Samples/BookForList.cs]: BookForList is not a single word moreover, what does it mean? If you mean List of Books then you should only names it BookList instead

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 2]
```diff
@@ -8,196 +8,196 @@
...
+                int previousPage;
+                int nextPage;
+                if(searchString == null)
                 {
-                    BookId = p.Id,
-                    BookName = p.BookName,
-                    BookType = p.BookType,
-                    Description = p.Description
-                }).AsQueryable();
+                    var source = _bookRepository.Select(p => new BookDto
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: Please use AutoMapper in this line, it is easier for managing Object Mapping. For more information you can visit this page: https://docs.abp.io/en/abp/latest/Object-To-Object-Mapping

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 3]
```diff
@@ -8,196 +8,196 @@
...
+                        BookName = p.BookName,
+                        BookType = p.BookType,
+                        Description = p.Description
+                    }).AsQueryable();
 
-                count = source.Count();
+                    count = source.Count();
 
-                currentPage = pageNumber;             
+                    currentPage = pageNumber;
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: Why not directly use pageNumber?

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 4]
```diff
@@ -8,196 +8,196 @@
...
 
-                totalCount = count;
+                    totalCount = count;
 
-                totalPage = (int)Math.Ceiling(count / (double)pageSize);
+                    totalPage = (int)Math.Ceiling(count / (double)pageSize);
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: Same here, why would you assign count to another variable(totalCount)?

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 5]
```diff
@@ -8,196 +8,196 @@
...
-                    PreviousPage = previousPage,
-                    NextPage = nextPage,
-                    Books = books
-                };
-                return result;
-            }
-            //count record
-            
+                    // source query with search string
+                    var source = _bookRepository.Where(p => p.BookName.Contains(searchString)).Select(p => new BookDto
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: there is a hefty code duplication here, you should refactor this method.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 6]
```diff
@@ -8,196 +8,196 @@
...
-            if (bookDetail == null)
+            public async Task<BookDto> GetBookById(Guid id)
             {
-                return null;
+                var bookDetail = await _bookRepository.FirstOrDefaultAsync(p => p.Id == id);
+                if (bookDetail == null)
+                {
+                    return null;
+                }
+                return new BookDto
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: You should use AutoMapper here.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 7]
```diff
@@ -8,196 +8,196 @@
...
-                Description = bookDetail.Description
-            };
-        }
 
-        public async Task<int> DeleteBook(Guid id)
-        {
-            var bookForDelete = await _bookRepository.FirstOrDefaultAsync(p => p.Id == id);
-            if (bookForDelete == null)
+            public async Task<int> DeleteBook(Guid id)
             {
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs:64]: Return true or false not 1 or 0, The returned Information is ambiguous if you persist this way.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 8]
```diff
@@ -8,196 +8,196 @@
...
-        public async Task<int> UpdateBook(Guid id, BookDto bookDto)
-        {
-            var existedBook = await _bookRepository.FirstOrDefaultAsync(b => b.Id == id);
-       
-            if (existedBook == null)
-            {
-                return 0;
-            }
-            else
+            public async Task<int> UpdateBook(Guid id, BookDto bookDto)
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: Same here, return true or false.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 9]
```diff
@@ -8,196 +8,196 @@
...
+                    return 1;
+                }
             }
-            return 0;
-        }
-        public async Task<int> DeteleBook(Guid id)
-        {
-            var bookToDetele =  await _bookRepository.FirstOrDefaultAsync(p => p.Id == id);
-            if(bookToDetele == null)
+            public async Task<int> CreateBook(BookDto bookDto)
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: Same here, return true or false.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs [hunk 10]
```diff
@@ -8,196 +8,196 @@
...
+                    book.BookType = bookDto.BookType;
+                    book.Description = bookDto.Description;
+                    await _bookRepository.InsertAsync(book);
+                    return 1;
+                }
                 return 0;
             }
-            else
+            public async Task<int> DeteleBook(Guid id)
             {
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Application/Samples/BookServices.cs]: Why does it has 2 Delete methods?

File: Acme.ProjectCompare/src/Acme.ProjectCompare.Domain/ProjectCompareDbProperties.cs [hunk 11]
```diff
@@ -1,12 +1,11 @@
-﻿namespace Acme.ProjectCompare
-{
-    public static class ProjectCompareDbProperties
+﻿    namespace Acme.ProjectCompare
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.Domain/ProjectCompareDbProperties.cs:1]: please rename the connection string and the table prefix.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi.Client/ProjectCompareHttpApiClientModule.cs [hunk 12]
```diff
@@ -2,22 +2,22 @@
...
-namespace Acme.ProjectCompare
-{
-    [DependsOn(
-        typeof(ProjectCompareApplicationContractsModule),
-        typeof(AbpHttpClientModule))]
-    public class ProjectCompareHttpApiClientModule : AbpModule
+    namespace Acme.ProjectCompare
     {
-        public const string RemoteServiceName = "ProjectCompare";
+        [DependsOn(
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi.Client/ProjectCompareHttpApiClientModule.cs:7]: Rename the remote service.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs [hunk 13]
```diff
@@ -6,65 +6,65 @@
...
-    [RemoteService]
-    [Route("api/ProjectCompare/book")]
-    public class BookController : ProjectCompareController
+    namespace Acme.ProjectCompare.Samples
     {
-        private readonly IBookServices _bookServices;
-        public BookController(IBookServices bookServices)
+        [RemoteService]
+        [Route("api/ProjectCompare/book")]
+        public class BookController : ProjectCompareController
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs]: Change the controller Route, "ProjectCompare" is not valid in your context.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs [hunk 14]
```diff
@@ -6,65 +6,65 @@
...
-            if(result == 0)
+                var result = await _bookServices.GetBooks(pageSize,pageNumber,searchString);       
+                return Json(result);
+            }
+            [HttpGet("{id}")]
+            public async Task<IActionResult> GetBookById(Guid id)
             {
-                return new BadRequestObjectResult(new { Message = "Update failed" });
+                var book = await _bookServices.GetBookById(id);
+                if(book == null)
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs]: The If statement is pointless.

File: Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs [hunk 15]
```diff
@@ -6,65 +6,65 @@
...
-        {
-            var result = await _bookServices.DeleteBook(id);
-            if (result == 0)
+        
+            [HttpPost]
+            public async Task<IActionResult> CreateBook([FromBody] BookDto bookDto)
             {
-                return new BadRequestObjectResult(new { Message = "Delele Error" });
+                var result = await _bookServices.CreateBook(bookDto);
+                if(result == 0)
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs]: compare true or false when the result comes back from the service layer. (not 1 or 0)

File: Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs [hunk 16]
```diff
@@ -6,65 +6,65 @@
...
+                {
+                    return new BadRequestObjectResult(new { Message = "Update failed" });
+                }   
+                return Ok(new { Message = "Created Successfully" });
             }
-            return Ok(new { Message = "Deleted Successfully" });
+            [HttpPut("{id}")]
+            public async Task<IActionResult> UpdateBook(Guid id, [FromBody] BookDto bookDto)
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs]: compare true or false when the result comes back from the service layer. (not 1 or 0)

File: Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs [hunk 17]
```diff
@@ -6,65 +6,65 @@
...
+                if(result == 0)
+                {
+                    return new BadRequestObjectResult(new { Message = "Update failed" });
+                }
+                return Ok(new { Message = "Updated Successfully" });
+            }
+            [HttpDelete("{id}")]
+            public async Task<IActionResult> DeleteBook(Guid id)
+            {
+                var result = await _bookServices.DeleteBook(id);
```
Comment (phongdamtps) [inline Acme.ProjectCompare/src/Acme.ProjectCompare.HttpApi/Samples/BookController.cs]: compare true or false when the result comes back from the service layer. (not 1 or 0)

File: BookAngular/src/app/book/listbook/listbook.component.ts [hunk 18]
```diff
@@ -13,105 +13,105 @@ import { BookServiceService } from '../../services/book-service.service';
...
-  
-  ngOnInit() {
-      this.loadBooks();
-      console.log("On init", this.isReloadBooks)
-      if(this.isReloadBooks)
-      {
+    // dialog field 
+    public isActive = false;
+    //check dialog create or edit
+    public isNew = false;
```
Comment (phongdamtps) [inline BookAngular/src/app/book/listbook/listbook.component.ts:26]: The variable name should imply the meaning of the subject description, you should rename the variable and remove the comment.

File: BookAngular/src/app/book/listbook/listbook.component.ts [hunk 19]
```diff
@@ -13,105 +13,105 @@ import { BookServiceService } from '../../services/book-service.service';
...
-  constructor(public service: BookServiceService
-    ) {}
-  
-  ngOnInit() {
-      this.loadBooks();
-      console.log("On init", this.isReloadBooks)
-      if(this.isReloadBooks)
-      {
+    // dialog field 
+    public isActive = false;
```
Comment (phongdamtps) [inline BookAngular/src/app/book/listbook/listbook.component.ts]: The variable name should imply the meaning of the subject description, you should rename the variable and remove the comment.

File: BookAngular/src/app/book/listbook/listbook.component.ts [hunk 20]
```diff
@@ -13,105 +13,105 @@ import { BookServiceService } from '../../services/book-service.service';
...
-      }
-  }
-
-  onPageChange(e: PageChangeEvent) {
-    this.pagingModel.skip = e.skip;
-    this.pagingModel.pageSize = e.take;
-    this.pagingModel.currentPage = Math.ceil((this.pagingModel.skip / this.pagingModel.pageSize) + 1)
-    this.loadBooks();
-  }
+        console.log("On init", this.isReloadBooks)
```
Comment (phongdamtps) [inline BookAngular/src/app/book/listbook/listbook.component.ts]: * Remove all the console logs in the source code before pushing.

File: BookAngular/src/app/book/listbook/listbook.component.ts [hunk 21]
```diff
@@ -13,105 +13,105 @@ import { BookServiceService } from '../../services/book-service.service';
...
+      this.isReloadBooks = true;
+    }
 
-  onEditBook() {
-    this.isActive = false;
-  }
+    cancelHandler(){
+      this.isActive = false;
+      this.isActiveDeleteDialog = false;
+      //Empty Data
```
Comment (phongdamtps) [inline BookAngular/src/app/book/listbook/listbook.component.ts]: remove comment

File: BookAngular/src/app/book/listbook/listbook.component.ts [hunk 22]
```diff
@@ -13,105 +13,105 @@ import { BookServiceService } from '../../services/book-service.service';
...
-    else if(searchString !== null){
-      getBookCallback = (this.pagingModel.skip === 0 && this.pagingModel.pageSize == 2 ) ? this.service.GetBook() : this.service.GetBook(this.pagingModel.pageSize, this.pagingModel.currentPage, this.searchString);
+
+    public loadBooks(searchString?: string) {
+      var getBookCallback;
+      if(searchString == null)
+      {
+        getBookCallback = (this.pagingModel.skip === 0 && this.pagingModel.pageSize == 2 ) ? this.service.GetBook() : this.service.GetBook(this.pagingModel.pageSize, this.pagingModel.currentPage);
+      }
+      else if(searchString !== null){
```
Comment (phongdamtps) [inline BookAngular/src/app/book/listbook/listbook.component.ts]: simply use else.

File: BookAngular/src/app/services/book-service.service.ts [hunk 23]
```diff
@@ -1,71 +1,71 @@
...
+          return result;
+        } else {
+          result = this.http.get(this.baseUrlPaging);
+          return result
+        }
+      }
+      //Search
+      else if(searchString != null){
+        if (pageSize !== undefined && pageNumber !== undefined) {      
+          const urlPaging =  `https://localhost:44317/api/ProjectCompare/book?pageSize=${pageSize}&pageNumber=${pageNumber}&searchString=${searchString}`
```
Comment (phongdamtps) [inline BookAngular/src/app/services/book-service.service.ts]: Use Url variable reference instead. Shouldn't set it statically like this.

Comment (notelevision99) [reply BookAngular/src/app/services/book-service.service.ts:34]: Can you explain more clearly this issue

**Pull request merged**